plugins/storage/volume/primera/src/main/java/org/apache/cloudstack/storage/datastore/adapter/primera/PrimeraAdapter.java

```diff
@@ -0,0 +1,911 @@
+
+        Long virtualSizeInBytes = vol.getHostWriteMiB()  * PrimeraAdapter.BYTES_IN_MiB;
+        Long allocatedSizeInBytes = vol.getSizeMiB() * PrimeraAdapter.BYTES_IN_MiB;
+        Long actualUsedInBytes = vol.getTotalUsedMiB() * PrimeraAdapter.BYTES_IN_MiB;
+        ProviderVolumeStats stats = new ProviderVolumeStats();
+        stats.setActualUsedInBytes(actualUsedInBytes);
+        stats.setAllocatedInBytes(allocatedSizeInBytes);
+        stats.setVirtualUsedInBytes(virtualSizeInBytes);
+        return stats;
+    }
+
+    @Override
+    public boolean canAccessHost(ProviderAdapterContext context, String hostname) {
+        PrimeraHostset hostset = getHostset(this.hostset);
+
+        List<String> members = hostset.getSetmembers();
+
+        // check for fqdn and shortname combinations.  this assumes there is at least a shortname match in both the storage array and cloudstack
+        // hostname configuration
+        String shortname;
+        if (hostname.indexOf('.') > 0) {
+            shortname = hostname.substring(0, (hostname.indexOf('.')));
+        } else {
+            shortname = hostname;
+        }
+        for (String member: members) {
+            // exact match (short or long names)
+            if (member.equals(hostname)) {
+                return true;
+            }
+
+            // primera has short name and cloudstack had long name
+            if (member.equals(shortname)) {
+                return true;
+            }
+
+            // member has long name but cloudstack had shortname
+            if (member.substring(0, (member.indexOf('.'))).equals(shortname)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private PrimeraCpg getCpg(String name) {
+        return GET("/cpgs/" + name, new TypeReference<PrimeraCpg>() {
+        });
+    }
+
+    private PrimeraHostset getHostset(String name) {
+        return GET("/hostsets/" + name, new TypeReference<PrimeraHostset>() {
+        });
+    }
+
+    private String getSessionKey() {
+        refreshSession(false);
+        return key;
+    }
+
+    private synchronized void refreshSession(boolean force) {
+        try {
+            if (force || keyExpiration < System.currentTimeMillis()) {
+                // close client to force connection reset on appliance -- not doing this can result in NotAuthorized error...guessing
+                _client.close();;
+                _client = null;
+                login();
+                keyExpiration = System.currentTimeMillis() + keyTtl;
+            }
+        } catch (Exception e) {
+            // retry frequently but not every request to avoid DDOS on storage API
+            logger.warn("Failed to refresh Primera API key for " + username + "@" + url + ", will retry in 5 seconds", e);
+            keyExpiration = System.currentTimeMillis() + (5*1000);
+        }
+    }
+    /**
+     * Login to the array and get an access token
+     */
+    private void login() {
```

Comment:
very long method, can it be partitioned?

Reply:
Updated in upcoming commit.